sway-ir/src/instruction.rs

```diff
@@ -341,6 +341,30 @@ impl Instruction {
             }
         }
     }
+
+    pub fn may_have_side_effect(&self) -> bool {
+        matches!(
+            self,
+            Instruction::AsmBlock(_, _)
+                | Instruction::Call(_, _)
+                | Instruction::ContractCall { .. }
+                | Instruction::StateLoadQuadWord { .. }
+                | Instruction::StateStoreQuadWord { .. }
+                | Instruction::StateStoreWord { .. }
+                | Instruction::Store { .. }
+                // Insert(Element/Value), unlike those in LLVM
+                // do not have SSA semantics. They are like stores.
```

Comment:
I don't think this is true, or at least it isn't supposed to be true.  Hrmm, at least semantically they return new values which are then used instead of their target arg values.

Reply:
That's what I thought so too first. But the `struct_field_access` test has IR where the result of this instruction isn't used at all (which I ended up DCE'ing, but that fails the test), and I then confirmed by looking at ASM gen that it works by assigning the same reg to the input and output operand of these instructions. In other words, the output doesn't need to be used at all, explicitly.

Reply:
`insert_value` kinda sucks and I've targeted fixing them (and aggregates in general) in the future IR enhancements... :slightly_frowning_face:

So OK, whatever works for now.

Reply:
```script {    
    fn main() -> u64, !1 {
        local mut ptr { u64 } data     
        local ptr u64 other
                                                  
        entry:
        v0 = get_ptr mut ptr { u64 } data, ptr { u64 }, 0, !2
        v1 = const { u64 } { u64 42 }, !3
        store v1, ptr v0, !2                                                                         
        v2 = get_ptr mut ptr { u64 } data, ptr { u64 }, 0, !4                                                                                                                                              
        v3 = const u64 43, !5
        v4 = insert_value v2, { u64 }, v3, 0, !4
        v5 = const { u64 } { u64 44 }, !6
        v6 = extract_value v5, { u64 }, 0, !7
        v7 = get_ptr ptr u64 other, ptr u64, 0, !8 
        store v6, ptr v7, !8
        v8 = get_ptr mut ptr { u64 } data, ptr { u64 }, 0, !9
        v9 = extract_value v8, { u64 }, 0, !7
        ret u64 v9, !10
    }

    fn ret_struct_0() -> { u64 }, !11 {
        entry:
        v0 = const { u64 } { u64 undef }, !6
        v1 = const u64 44, !12
        v2 = insert_value v0, { u64 }, v1, 0, !6
        ret { u64 } v2
    }
}
```

Here is the IR generated for `struct_field_access`. Here `v4 = insert_value v2, { u64 }, v3, 0, !4` isn't dead. Removing it leads to incorrect assembly generated.

Reply:
It's not dead, but it isn't producing a side-effect like `store` as the comment suggests..?

Reply:
> It's not dead, but it isn't producing a side-effect like `store` as the comment suggests..?

If the value that it's defining `v4`, (which is the instruction itself) isn't being used, but the instruction still is required for correctness, then I'd say that it has a side effect.

Reply:
Having said that, I can modify or remove the comment as you think it should be. I just wrote what I understood about how it's working (which could be wrong).

Reply:
Yep, I see now, it's pretty broken and the aggregates refactor is overdue.  (While still not being super urgent.)